ambry-store/src/main/java/com/github/ambry/store/IndexSegment.java

```diff
@@ -1185,33 +1233,173 @@ static Offset getIndexSegmentStartOffset(String filename) {
+      if (cursor < 0 || cursor > numberOfEntries) {
+        throw new NoSuchElementException();
+      }
       try {
-        StoreKey key = getKeyAt(mmap, currentIdx);
+        StoreKey key = getKeyAt(mmap, cursor);
         mmap.get(valueBuf);
         return new IndexEntry(key, new IndexValue(startOffset.getName(), ByteBuffer.wrap(valueBuf), getVersion()));
       } catch (Exception e) {
-        String message = "Failed to read index entry at " + currentIdx;
+        String message = "Failed to read index entry at " + cursor;
         logger.error(message, e);
         throw new IllegalStateException(message, e);
       } finally {
-        currentIdx++;
+        cursor++;
+      }
+    }
+  }
+
+  /**
+   * An {@link IndexEntry} {@link ListIterator} for a sealed {@link IndexSegment}.
+   */
+  private class SealedIndexSegmentEntryListIterator extends SealedIndexSegmentEntryIterator
+      implements ListIterator<IndexEntry> {
+
+    SealedIndexSegmentEntryListIterator(int currentIndex) {
+      this.cursor = currentIndex;
+    }
+
+    @Override
+    public boolean hasPrevious() {
+      return cursor != 0;
+    }
+
+    @Override
+    public IndexEntry previous() {
+      try {
+        int i = cursor - 1;
+        StoreKey key = getKeyAt(mmap, i);
+        mmap.get(valueBuf);
+        return new IndexEntry(key, new IndexValue(startOffset.getName(), ByteBuffer.wrap(valueBuf), getVersion()));
+      } catch (Exception e) {
+        String message = "Failed to read index entry at " + cursor;
+        logger.error(message, e);
+        throw new IllegalStateException(message, e);
+      } finally {
+        cursor--;
+      }
+    }
+
+    @Override
+    public int nextIndex() {
+      return cursor;
+    }
+
+    @Override
+    public int previousIndex() {
+      return cursor - 1;
+    }
+
+    @Override
+    public void remove() {
+      throw new UnsupportedOperationException("Remove unsupported");
+    }
+
+    @Override
+    public void set(IndexEntry indexEntry) {
+      throw new UnsupportedOperationException("Set unsupported");
+    }
+
+    @Override
+    public void add(IndexEntry indexEntry) {
+      throw new UnsupportedOperationException("Add unsupported");
+    }
+  }
+
+  /**
+   * An {@link IndexEntry} {@link Iterator} for a unsealed {@link IndexSegment}.
+   */
+  private class UnsealedIndexSegmentEntryIterator implements Iterator<IndexEntry> {
+    protected final ArrayList<IndexEntry> entries = new ArrayList<>();
+    protected final Iterator<IndexEntry> it;
+
+    UnsealedIndexSegmentEntryIterator() {
+      NavigableMap<StoreKey, ConcurrentSkipListSet<IndexValue>> indexMap = index;
+      StoreKey keyCursor = indexMap.firstKey();
+      while (keyCursor != null) {
+        ConcurrentSkipListSet<IndexValue> indexValues = indexMap.get(keyCursor);
+        Iterator<IndexValue> it = indexValues.iterator();
+        while (it.hasNext()) {
+          entries.add(new IndexEntry(keyCursor, it.next()));
```

Comment:
Is the purpose of copying the entries here to create a snapshot and avoid issues with the unsealed segment becoming sealed while iterating through it?

Reply:
not really, we can keep a reference to the index map in the iterator so that even if it becomes sealed, we still have all the data we need in the memory. The reason really is for listiterator. Iterator iterate through entries from the smallest to the largest ( ordered by key then by offset), but what we need in the blobstorestats is a ListIterator that can iterate through entries from the largest to the smallest. There are ways to create iterator without making an arraylist of entries, but this is by far the easiest way. And it does take up too much memory since there is no copy of IndexEntry while creating this list

Reply:
agreed that this is probably the easiest way for now. I guess if we specifically wanted a reverse iterator (and not a general listiterator), you may be able to use`index.descendingMap()` but we don't have to go this route unless we see a perf problem with the current approach.